pkg/agent/cluster.go

```diff
@@ -156,60 +157,79 @@ func (czero *Cluster) IsBootstrapComplete() (bool, error) {
 		if !czero.installHistory.RestAPISeen {
 			logrus.Debug("Agent Rest API Initialized")
 			czero.installHistory.RestAPISeen = true
+			czero.installHistory.NotReadyTime = time.Now()
 		}
 
 		// Lazy loading of the clusterID and clusterInfraEnvID
 		if czero.clusterID == nil {
 			clusterID, err := czero.API.Rest.getClusterID()
 			if err != nil {
-				return false, errors.Wrap(err, "Unable to retrieve clusterID from Agent Rest API")
+				return false, false, errors.Wrap(err, "Unable to retrieve clusterID from Agent Rest API")
 			}
 			czero.clusterID = clusterID
 		}
 
 		if czero.clusterInfraEnvID == nil {
 			clusterInfraEnvID, err := czero.API.Rest.getClusterInfraEnvID()
 			if err != nil {
-				return false, errors.Wrap(err, "Unable to retrieve clusterInfraEnvID from Agent Rest API")
+				return false, false, errors.Wrap(err, "Unable to retrieve clusterInfraEnvID from Agent Rest API")
 			}
 			czero.clusterInfraEnvID = clusterInfraEnvID
 		}
 
 		logrus.Trace("Getting cluster metadata from Agent Rest API")
 		clusterMetadata, err := czero.GetClusterRestAPIMetadata()
 		if err != nil {
-			return false, errors.Wrap(err, "Unable to retrieve cluster metadata from Agent Rest API")
+			return false, false, errors.Wrap(err, "Unable to retrieve cluster metadata from Agent Rest API")
 		}
 
 		if clusterMetadata == nil {
-			return false, errors.New("cluster metadata returned nil from Agent Rest API")
+			return false, false, errors.New("cluster metadata returned nil from Agent Rest API")
 		}
 
-		if !checkHostsValidations(clusterMetadata, logrus.StandardLogger()) {
-			return false, errors.New("cluster host validations failed")
+		czero.PrintInstallStatus(clusterMetadata)
+
+		if *clusterMetadata.Status == models.ClusterStatusReady {
+			// If the status changes back to Ready from Installing it indicates an error
+			if czero.installHistory.RestAPIPreviousClusterStatus == models.ClusterStatusPreparingForInstallation {
+				return false, true, errors.New("failed to prepare cluster installation")
+			}
+
+			// Check if stuck in Ready state
+			if czero.installHistory.RestAPIPreviousClusterStatus == models.ClusterStatusReady {
+				current := time.Now()
+				elapsed := current.Sub(czero.installHistory.NotReadyTime)
+				if elapsed > 1*time.Minute {
+					return false, true, errors.New("failed to progress after all hosts available")
+				}
+			}
+		} else {
+			czero.installHistory.NotReadyTime = time.Now()
 		}
```

Comment:
Lets move this to its own function instead of adding a second boolean to the overall function. The second boolean is confusing to me coming out of `IsBootstrapComplete()`. Maybe something like `RegressionInClusterStatus()`

Reply:
Moved to function